diff_parallel.go

```diff
@@ -0,0 +1,363 @@
+package hamt
+
+import (
+	"bytes"
+	"context"
+	"sync"
+
+	"github.com/ipfs/go-cid"
+	cbor "github.com/ipfs/go-ipld-cbor"
+	cbg "github.com/whyrusleeping/cbor-gen"
+	"golang.org/x/sync/errgroup"
+	"golang.org/x/sync/semaphore"
+	"golang.org/x/xerrors"
+)
+
+// ParallelDiff returns a set of changes that transform node 'prev' into node 'cur'. opts are applied to both prev and cur.
+func ParallelDiff(ctx context.Context, prevBs, curBs cbor.IpldStore, prev, cur cid.Cid, workers int64, opts ...Option) ([]*Change, error) {
+	if prev.Equals(cur) {
+		return nil, nil
+	}
+
+	prevHamt, err := LoadNode(ctx, prevBs, prev, opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	curHamt, err := LoadNode(ctx, curBs, cur, opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	if curHamt.bitWidth != prevHamt.bitWidth {
+		return nil, xerrors.Errorf("diffing HAMTs with differing bitWidths not supported (prev=%d, cur=%d)", prevHamt.bitWidth, curHamt.bitWidth)
+	}
+
+	return doParallelDiffNode(ctx, prevHamt, curHamt, workers)
+}
+
+func doParallelDiffNode(ctx context.Context, pre, cur *Node, workers int64) ([]*Change, error) {
+	bp := cur.Bitfield.BitLen()
+	if pre.Bitfield.BitLen() > bp {
+		bp = pre.Bitfield.BitLen()
+	}
+
+	initTasks := []*task{}
+	for idx := bp; idx >= 0; idx-- {
+		preBit := pre.Bitfield.Bit(idx)
+		curBit := cur.Bitfield.Bit(idx)
+		initTasks = append(initTasks, &task{
+			idx:    idx,
+			pre:    pre,
+			preBit: preBit,
+			cur:    cur,
+			curBit: curBit,
+		})
+	}
+
+	out := make(chan *Change, 2*workers)
+	differ, ctx := newDiffScheduler(ctx, workers, initTasks...)
+	differ.startScheduler(ctx)
+	differ.startProcessor(ctx, out)
+
+	var changes []*Change
+	done := make(chan struct{})
+	go func() {
+		for change := range out {
+			changes = append(changes, change)
+		}
+		close(done)
+	}()
+
+	if err := differ.grp.Wait(); err != nil {
+		close(out)
+		return nil, err
+	}
+	close(out)
+	<-done
+
+	return changes, nil
```

Comment:
```suggestion
	err := differ.grp.Wait()
	close(out)
	<-done

	return changes, err
```

It's the "nicer" thing to do.